Fixes part of #6240:  Mark correct answers after enabling correctness feedback

cuichenli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR adds e2e tests for mark correct answer after enabling correctness feedback. 

Other changes:
- Add one function `setResponse`. This is used to set the response when the `Add response` modal is opened.
- Change `getResponseEditor` to be able to return the currently opened `Add response modal`. 
- Add `waitForClickable` in `createNewInteraction`. Due to a strange behaviour: if I set the interaction to be other than general ones, e.g. NumericInput, the test will run to `interactionTile(interactionId).click();` and then complain the tile is not displayed (maybe due to I was using another version of chrome driver). 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

cuichenli committed:
Draft version

cuichenli committed:
Draft version for mark correctness  e2e test

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Another test

cuichenli committed:
Add e2e test for enable correctness feedback.

Draft

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Fix addResponse.apply

DubeySandeep commented on core/tests/protractor.conf.js:
```diff
@@ -44,6 +44,10 @@ var suites = {
       'protractor_desktop/explorationFeedbackTab.js'
     ],
+  
+    explorationEnableCorrectnessFeedback: [
```
Let's not create a new suite, can we add this to the coreEditorAndPlayerFeatures suite. or maybe we can create an explorationSettingsTab suite?

nithusha21 replied:
I think coreEditorAndPlayerFeatures sounds good. Can we do that?

cuichenli replied:
Updated in the new commit

DubeySandeep commented on core/tests/protractor_desktop/explorationEnableCorrectnessFeedback.js:
```diff
@@ -0,0 +1,142 @@
+  var explorationEditorSettingsTab = null;
+
+  var tickMark = element(
+    by.css('.protractor-test-correct-tick-mark'));
```
Move selectors and functions to the utils file*

cuichenli replied:
Update

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorSettingsTab.js:
```diff
@@ -105,6 +107,10 @@ var ExplorationEditorSettingsTab = function() {
   };
 
+  this.enableCorrectnessFeedback = function() {
+    enableCorrectnessFeedbackButton.click();
```
Add a check to make sure this element is visible and clickable*

cuichenli replied:
Update

DubeySandeep replied:
Add `waitFor.elementToBeClickable(` to check whether it's clickable.

cuichenli replied:
Added

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -291,24 +295,26 @@ var ExplorationEditorMainTab = function() {
-        headerElem.click();
+    if (responseNum !== 'pop') {
+      var headerElem;
+      if (responseNum === 'default') {
```
Can you explain the reason for this change?

cuichenli replied:
This aims to get the opened `Add Response` modal during `setInteraction`. Usually, it is immediately closed after the interaction is set. 
https://github.com/oppia/oppia/blob/f99921188031c10c6609ef72d37e3a67de297a15/core/tests/protractor_utils/ExplorationEditorMainTab.js#L623

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -291,24 +295,26 @@ var ExplorationEditorMainTab = function() {
 
+      responseBody(responseNum).isPresent().then(function(isVisible) {
+        if (!isVisible) {
+          expect(headerElem.isDisplayed()).toBe(true);
```
For `responseNum === 'pop'` what will be `headerElem`?

cuichenli replied:
These lines will not be executed.  It jumps to the return statement directly.

DubeySandeep replied:
I'm not sure whether `headerElem` will be defined for `responseNum === 'pop'`?

cuichenli replied:
No, it would not.
But I have added one line to set `headerElem` to `addResponseHeader` and wait for it to be displayed.

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -657,7 +667,13 @@ var ExplorationEditorMainTab = function() {
+      targetTile,
+      'Interaction tile ' + interactionId + ' takes too long to be clickable'
+    );
+    expect(targetTile.isDisplayed()).toBe(true);
```
Should we firstly check it's visible then clickable?

cuichenli replied:
Update

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -657,7 +667,13 @@ var ExplorationEditorMainTab = function() {
+      'Interaction tile ' + interactionId + ' takes too long to be clickable'
+    );
+    expect(targetTile.isDisplayed()).toBe(true);
     interactionTile(interactionId).click();
```
Can we use targetTile here?

cuichenli replied:
Updated

DubeySandeep commented on core/tests/protractor_desktop/explorationEnableCorrectnessFeedback.js:
```diff
@@ -0,0 +1,142 @@
+  });
+
+  it('should enable correctness feedback after the interaction is created.' +
+     'Correctness feedback is set in response editor.', function() {
```
`should allow selecting correct feedback from the response editor after the interaction is created`

cuichenli replied:
Updated

DubeySandeep commented on core/tests/protractor_desktop/explorationEnableCorrectnessFeedback.js:
```diff
@@ -0,0 +1,142 @@
+  });
+
+  it('should turn on enable correctness feedback before create interaction.' +
+     'Correctness feedback is set in add response modal.', function() {
```
Make it in a single sentence.

cuichenli replied:
Updated

DubeySandeep commented on core/tests/protractor_desktop/explorationEnableCorrectnessFeedback.js:
```diff
@@ -0,0 +1,142 @@
+  });
+
+  it('should turn on correctness feedback before create interaction.' +
+    'Correctness feedback is set in response editor.', function() {
```
Ditto*

cuichenli replied:
Updated

DubeySandeep commented on core/tests/protractor_desktop/explorationEnableCorrectnessFeedback.js:
```diff
@@ -0,0 +1,142 @@
+    explorationEditorMainTab.moveToState('End');
+    explorationEditorMainTab.setInteraction('EndExploration');
+    explorationEditorPage.saveChanges();
+  });
```
Should we also test the expected behavior of these changes to get reflected in the player mode? (OR that's going to be a separate test?)

cuichenli replied:
Added in afterEach

DubeySandeep commented on core/tests/protractor_desktop/extensions.js:
```diff
@@ -134,7 +134,7 @@ describe('Interactions', function() {
           null, [interactionId].concat(test.interactionArguments));
 
-        explorationEditorMainTab.addResponse.apply(null, [
+        explorationEditorMainTab.addResponse.apply(explorationEditorMainTab, [
```
Can you explain the reason for this change?

cuichenli replied:
If the `explorationEditorMainTab` is not passed in, we will get `undefined` error https://travis-ci.org/oppia/oppia/jobs/596916385#L4852. 
I slightly changed the code in `addResponse` and if no context is passed in, this line will complain. 
https://github.com/oppia/oppia/blob/f99921188031c10c6609ef72d37e3a67de297a15/core/tests/protractor_utils/ExplorationEditorMainTab.js#L256

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -657,7 +667,13 @@ var ExplorationEditorMainTab = function() {
+
+    var targetTile = interactionTile(interactionId);
+    waitFor.elementToBeClickable(
+      targetTile,
```
is this resolve the issue you have mentioned in the PR explanation?

cuichenli replied:
Yes, this solved the issue.

DubeySandeep reviewed (commented):
@cuichenli, Thanks for the PR, I took a pass and left a few comments PTAL!

nithusha21 reviewed (commented):
I didn't do a full pass, will do so once @DubeySandeep's review is addressed. Thanks @cuichenli for the PR, and thanks @DubeySandeep for doing a review :)

cuichenli committed:
Add e2e test for exploration players with correctness feedback enabled

cuichenli committed:
Add more conditions and restructure

cuichenli committed:
Change it statements

cuichenli committed:
Move enable correctness e2e test into core editor features

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

DubeySandeep commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -144,6 +144,12 @@ var ExplorationEditorMainTab = function() {
+  /*
+   * Symbols
+   */
+  var tickMark = element(
```
Rename this to correctAnswerCheckbox (or something around that)

cuichenli replied:
Updated

DubeySandeep commented on core/tests/protractor_utils/ExplorationPlayerPage.js:
```diff
@@ -71,6 +71,8 @@ var ExplorationPlayerPage = function() {
     by.css('.protractor-test-return-to-parent-button'));
   var answerDetailsSubmitButton = element(
     by.css('.protractor-test-answer-details-submit-button'));
+  var correctFeedback = element(
```
Make it correctnessFeedbackElement*

cuichenli replied:
Updated

DubeySandeep reviewed (commented):
@cuichenli, I've left a few follow-up comments PTAL!

@nithusha21, can you please review this PR as I won't be very active for a few days and won't be able to review this PR quickly!

cuichenli committed:
Merge branch 'e2e-exploration-mark-correct' of github.com:cuichenli/oppia into e2e-exploration-mark-correct

cuichenli committed:
Rename elements and add some logic

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Fix lint issues

cuichenli committed:
Remove mistakenly added file

DubeySandeep commented on core/tests/protractor.conf.js:
```diff
@@ -44,6 +44,10 @@ var suites = {
       'protractor_desktop/explorationFeedbackTab.js'
     ],
+    
+    explorationMarkCorrect: [
```
Is this still required? I think all the e2e tests are failing because of this PTAL!

cuichenli replied:
Removed

DubeySandeep commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -33,6 +33,148 @@ var ExplorationEditorPage =
+  ];
+
+
+  var applyCommonSettings = function() {
```
Should we move this to the beforeEach section after creating an exploration? I think this will make the tests more readable.

(If required we can explicitly call `enableCorrectnessFeedback` whenever required in the test.)

nithusha21 replied:
I think moving it to the before Each is a good idea..

DubeySandeep commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -33,6 +33,148 @@ var ExplorationEditorPage =
+      null, correctOptions[currentExplorationIndex]);
+    explorationPlayerPage.expectCorrectFeedback();
+    explorationPlayerPage.clickThroughToNextCard();
+    explorationPlayerPage.expectExplorationToBeOver();
```
I think the testing should be a part of `it` block, and beforeEach and afterEach should be for setup and cleanup respectively, so maybe we should move the tests (or the expected behavior of the test) in the `it` block!)

@nithusha21 do you have any thoughts on this? Should we define a method on top to reduce duplicate code and use that method in `it` block? OR it's fine to have these test over here?

nithusha21 replied:
@cuichenli I think a solution similar to applyCommonSettings would work here. Create a function and call them within the it blocks. Typically the complete test is present in the it() block, even if it is repetitive. In the afterEach block make sure that there are no console errors (You can see all other afterEach blocks follow that).

cuichenli replied:
I have moved some lines to `beforeEach` while leaving a function `enableMarkCorrectness` to just enable the option.

DubeySandeep reviewed (approved):
@cuichenli, I've left a few comments PTAL!

(Other than that it LGTM! [Approved]) 

@nithusha21, Can you please review this PR?

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -33,6 +33,148 @@ var ExplorationEditorPage =
+    ['TextInput', 'One'],
+    ['NumericInput', 3]
+  ];
+
```
Remove the extra newline.

cuichenli replied:
Updated

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -33,6 +33,148 @@ var ExplorationEditorPage =
+    explorationEditorMainTab.setContent(forms.toRichText(
+      'Select the right option.'));
+
+    // Create interaction without close the add response modal. Set
```
nit: without closing?

cuichenli replied:
Updated

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -33,6 +33,148 @@ var ExplorationEditorPage =
+    workflow.createExploration();
+  });
+
+  it('should allow selecting correct feedback from the response editor ' +
```
Reading this test suite I think it is ok to keep the applyCommonSettings function separate. You may ignore the above comment to move it into beforeEach().

nithusha21 reviewed (approved):
Hi @cuichenli, I took a pass. PTAL at the comments. Let me know if you would like to discuss any of these points!

nithusha21 reviewed (changes requested):
Oops approved accidentally! Will approve once the above review is addressed

cuichenli committed:
Reorganize after each and before each

cuichenli committed:
Remove redundtant suite

cuichenli committed:
Merge branch 'develop' of github.com:cuichenli/oppia into e2e-exploration-mark-correct

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Rename applyCommonSetitngs to enableMarkCorrectness

nithusha21 commented on core/tests/protractor.conf.js:
```diff
@@ -44,7 +44,7 @@ var suites = {
       'protractor_desktop/explorationFeedbackTab.js'
     ],
-
+    
```
This is a stray change. Please remove it

cuichenli replied:
Removed.

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -25,14 +25,164 @@ var users = require('../protractor_utils/users.js');
+      testEnableCorrectnessInPlayerPage();
+    });
+
+  afterEach(function() {
```
After each blocks should also assert that there are no console errors. Could you take a look at some of the existing test suites which do this check and replicate the same here?

DubeySandeep replied:
[Adding refrence]  In afterEach block assert that there are no console errors as suggested here: https://github.com/oppia/oppia/wiki/End-to-End-Tests#important-gotchas

cuichenli replied:
Did not notice that. Updated, thanks!

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -25,14 +25,164 @@ var users = require('../protractor_utils/users.js');
 var ExplorationPlayerPage =
   require('../protractor_utils/ExplorationPlayerPage.js');
 var LibraryPage = require('../protractor_utils/LibraryPage.js');
+var ExplorationEditorPage =
```
Put these dependencies in alphabetical order

cuichenli replied:
Updated, thanks!

nithusha21 commented on core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js:
```diff
@@ -25,14 +25,164 @@ var users = require('../protractor_utils/users.js');
+    ['NumericInput', 3]
+  ];
+
+  var enableMarkCorrectness = function() {
```
Maybe a better name would be enableCorrectnessFeedbackSetting.

cuichenli replied:
Updated, thanks!

nithusha21 reviewed (commented):
Looks much better. Thanks @cuichenli. I left a few comments.

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into e2e-exploration-mark-correct

cuichenli committed:
Fix issues

- Remove redundant spaces
- Remove redundant import
- Reorder import statements
- Rename utility function

nithusha21 reviewed (approved):
LGTM thanks @cuichenli!

nithusha21 commented:
@kevinlee12 @aks681 codeowner reviews please!

kevinlee12 reviewed (approved):
lgtm!

aks681 reviewed (approved):
lgtm as codeowner.
